relay-redis/src/real.rs

```diff
@@ -44,43 +58,105 @@ impl ConnectionLike for Connection<'_> {
+            }
             ConnectionInner::Single(ref mut con) => con.req_packed_commands(cmd, offset, count),
         }
     }
 
     fn get_db(&self) -> i64 {
-        match self.inner {
+        match self {
             ConnectionInner::Cluster(ref con) => con.get_db(),
+            ConnectionInner::MultiWrite {
+                primary: primary_connection,
+                ..
+            } => primary_connection.get_db(),
             ConnectionInner::Single(ref con) => con.get_db(),
         }
     }
 
     fn check_connection(&mut self) -> bool {
-        match self.inner {
+        match self {
             ConnectionInner::Cluster(ref mut con) => con.check_connection(),
+            ConnectionInner::MultiWrite {
+                primary: primary_connection,
+                secondaries: secondary_connections,
+            } => {
+                primary_connection.check_connection()
+                    && secondary_connections
+                        .iter_mut()
+                        .all(|c| c.check_connection())
+            }
             ConnectionInner::Single(ref mut con) => con.check_connection(),
         }
     }
 
     fn is_open(&self) -> bool {
-        match self.inner {
+        match self {
             ConnectionInner::Cluster(ref con) => con.is_open(),
+            ConnectionInner::MultiWrite {
+                primary: primary_connection,
+                secondaries: secondary_connections,
+            } => primary_connection.is_open() && secondary_connections.iter().all(|c| c.is_open()),
             ConnectionInner::Single(ref con) => con.is_open(),
         }
     }
 }
 
-enum PooledClientInner {
-    Cluster(Box<PooledConnection<redis::cluster::ClusterClient>>),
-    Single(Box<PooledConnection<redis::Client>>),
+/// A reference to a pooled connection from `PooledClient`.
+pub struct Connection<'a> {
+    inner: ConnectionInner<'a>,
+}
+
+impl ConnectionLike for Connection<'_> {
+    fn req_packed_command(&mut self, cmd: &[u8]) -> redis::RedisResult<redis::Value> {
+        self.inner.req_packed_command(cmd)
+    }
+
+    fn req_packed_commands(
+        &mut self,
+        cmd: &[u8],
+        offset: usize,
+        count: usize,
+    ) -> redis::RedisResult<Vec<redis::Value>> {
+        self.inner.req_packed_commands(cmd, offset, count)
+    }
+
+    fn get_db(&self) -> i64 {
+        self.inner.get_db()
+    }
+
+    fn check_connection(&mut self) -> bool {
+        self.inner.check_connection()
+    }
+
+    fn is_open(&self) -> bool {
+        self.inner.is_open()
+    }
 }
 
 /// A pooled Redis client.
-pub struct PooledClient {
-    opts: RedisConfigOptions,
-    inner: PooledClientInner,
+pub enum PooledClient {
+    /// Pool that is connected to a Redis cluster.
+    Cluster(
+        Box<PooledConnection<redis::cluster::ClusterClient>>,
+        RedisConfigOptions,
+    ),
+    /// Multiple pools that are used for multi-write.
+    MultiWrite(Box<PooledClient>, Vec<PooledClient>),
```

Comment:
We could also name primary/secondaries here like in `ConnectionInner`, but feel free to ignore if you think this is less readable in the using code.